packages/solana/src/providers/AuthProvider.ts

```diff
@@ -0,0 +1,203 @@
+      method: 'solana_signTransaction',
+      params: { transaction: this.serializeTransaction(transaction) }
+    })
+
+    const decodedTransaction = base58.decode(result.transaction)
+
+    if (isVersionedTransaction(transaction)) {
+      return VersionedTransaction.deserialize(decodedTransaction) as T
+    }
+
+    return Transaction.from(decodedTransaction) as T
+  }
+
+  public async signAndSendTransaction<T extends AnyTransaction>(
+    transaction: T,
+    options?: SendOptions
+  ) {
+    const serializedTransaction = this.serializeTransaction(transaction)
+
+    const result = await this.provider.request({
+      method: 'solana_signAndSendTransaction',
+      params: {
+        transaction: serializedTransaction,
+        options
+      }
+    })
+
+    return result.signature
+  }
+
+  public async sendTransaction(
+    transaction: AnyTransaction,
+    connection: Connection,
+    options?: SendOptions
+  ) {
+    const signedTransaction = await this.signTransaction(transaction)
+    const signature = await connection.sendRawTransaction(signedTransaction.serialize(), options)
+
+    return signature
+  }
+
+  // -- W3mFrameProvider methods ------------------------------------------- //
+  connectEmail: ProviderAuthMethods['connectEmail'] = args => this.provider.connectEmail(args)
+  connectOtp: ProviderAuthMethods['connectOtp'] = args => this.provider.connectOtp(args)
+  updateEmail: ProviderAuthMethods['updateEmail'] = args => this.provider.updateEmail(args)
+  updateEmailPrimaryOtp: ProviderAuthMethods['updateEmailPrimaryOtp'] = args =>
+    this.provider.updateEmailPrimaryOtp(args)
+  updateEmailSecondaryOtp: ProviderAuthMethods['updateEmailSecondaryOtp'] = args =>
+    this.provider.updateEmailSecondaryOtp(args)
+  getEmail: ProviderAuthMethods['getEmail'] = () => this.provider.getEmail()
+  getSocialRedirectUri: ProviderAuthMethods['getSocialRedirectUri'] = args =>
+    this.provider.getSocialRedirectUri(args)
+  connectDevice: ProviderAuthMethods['connectDevice'] = () => this.provider.connectDevice()
+  connectSocial: ProviderAuthMethods['connectSocial'] = args => this.provider.connectSocial(args)
+  connectFarcaster: ProviderAuthMethods['connectFarcaster'] = () => this.provider.connectFarcaster()
+  getFarcasterUri: ProviderAuthMethods['getFarcasterUri'] = () => this.provider.getFarcasterUri()
+  syncTheme: ProviderAuthMethods['syncTheme'] = args => this.provider.syncTheme(args)
+  syncDappData: ProviderAuthMethods['syncDappData'] = args => this.provider.syncDappData(args)
+  switchNetwork: ProviderAuthMethods['switchNetwork'] = async args => {
+    const result = await this.provider.switchNetwork(args)
+    this.emit('chainChanged', args as string)
+
+    return result
+  }
+
+  // -- Private ------------------------------------------- //
+  private getPublicKey<Required extends boolean>(
+    required?: Required
+  ): Required extends true ? PublicKey : PublicKey | undefined {
```

Comment:
what is this response type?

Reply:
if `required` is true then the function must return the `PublicKey` type, else it can return `PublicKey | undefined`